compatibility/sandbox-migration/SandboxMigrationRunner.hs

```diff
@@ -0,0 +1,177 @@
+            ]
+        Just Result{..} <- A.decodeFileStrict' outputFile
+        -- Test that all proposals are archived.
+        unless (null oldTProposals) $
+            fail ("Expected no old TProposals but got " <> show oldTProposals)
+        unless (null newTProposals) $
+            fail ("Expected no new TProposals but got " <> show newTProposals)
+        unless (prevTs == oldTs) $
+            fail ("Active ts should not have changed after migration: " <> show (prevTs, oldTs))
+        -- Test that no T contracts got archived.
+        let missingTs = filter (`notElem` newTs) oldTs
+        unless (null missingTs) $
+            fail ("The following contracts got lost during the migration: " <> show missingTs)
+        -- Test that only one new T contract is not archived.
+        let addedTs = filter (`notElem` oldTs) newTs
+        case addedTs of
+            [Tuple2 _ t] -> do
+                let expected = T testProposer testAccepter note
+                unless (t == expected) $
+                    fail ("Expected " <> show expected <> " but got " <> show t)
+            _ -> fail ("Expected 1 new T contract but got: " <> show addedTs)
+        hPutStrLn stderr ("<-- Tested " <> note)
+        pure newTs
+
+testProposer :: Party
+testProposer = Party "proposer"
+
+testAccepter :: Party
+testAccepter = Party "accepter"
+
+-- The datatypes are defined such that the autoderived Aeson instances
+-- match the DAML-LF JSON encoding.
+
+newtype ContractId t = ContractId T.Text
+  deriving newtype A.FromJSON
+  deriving stock (Eq, Show)
+newtype Party = Party T.Text
+  deriving newtype (A.FromJSON, A.ToJSON)
+  deriving stock (Eq, Show)
+
+data T = T
+  { proposer :: Party
+  , accepter :: Party
+  , note :: String
+  } deriving (Eq, Generic, Show)
+
+instance A.FromJSON T
+
+data TProposal = TProposal
+  { proposer :: Party
+  , accepter :: Party
+  , note :: T.Text
+  } deriving (Generic, Show)
+
+instance A.FromJSON TProposal
+
+data ScriptInput = ScriptInput
+  { _1 :: Party
+  , _2 :: Party
+  , _3 :: String
+  } deriving Generic
+
+instance A.ToJSON ScriptInput
+
+data Tuple2 a b = Tuple2
+  { _1 :: a
+  , _2 :: b
+  } deriving (Eq, Generic, Show)
+
+instance (A.FromJSON a, A.FromJSON b) => A.FromJSON (Tuple2 a b)
+
+data Result = Result
+  { oldTProposals :: [Tuple2 (ContractId TProposal) TProposal]
+  , newTProposals :: [Tuple2 (ContractId TProposal) TProposal]
+  , oldTs :: [Tuple2 (ContractId T) T]
+  , newTs :: [Tuple2 (ContractId T) T]
+  } deriving Generic
+
+instance A.FromJSON Result
+
+withSandbox :: FilePath -> T.Text -> (Int -> IO a) -> IO a
+withSandbox assistant jdbcUrl f =
+    withTempFile $ \portFile ->
+    withCreateProcess (proc assistant (args portFile)) { create_group = True } $ \_ _ _ ph -> do
+        p <- readPortFile maxRetries portFile
+        r <- f p
+        -- This is a shell process so we need to kill the whole process group.
+        interruptProcessGroupOf ph
```

Comment:
I haven’t looked too closely so far but I expect we probably have to fix this in the ledger tests as well. I’ll leave that for a separate PR.

Reply:
I ended up fixing the `Sandbox` module in this PR and switched over to this.